Aleksandr Nam contest4 task A

seshWCS opened the pull request.

seshWCS committed:
Create Set.h

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+        updh(root);
+        updh(ans);
+        return ans;
+    }
```
А обновить root->pr?

seshWCS replied:
он обновится в updLR(ans)

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+        if (v < root->val) {
+            root->l = insert(root->l, v);
+            root->l->pr = root;
+        } else if (root->val < v) {
```
А если равны?

seshWCS replied:
Ну я добавил иф, но он был опционален

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+        }
+    }
+
+    node* erasemin(node* root) {
```
Кажется по задумке функция должна разлинковать минимум в поддереве и вернуть его.
Тут происходит вообще что-то странное, разлинковки нет

seshWCS replied:
Нет, по задумке функция должна удалить минимум в поддереве root и вернуть оставшееся дерево после переподвешивания и ребалансировки.

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+                return a;
+            }
+            node* min = findmin(b);
+            min->r = erasemin(b);
```
тут в некоторых случаях могут потерятся целые поддеревья у b

seshWCS replied:
Нет, не могут, у меня же при удалении все перевешивается. Там же erasemin().

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+        if (nd->r != nullptr) {
+            destruct_node(nd->r);
+        }
+        if (nd->pr != nullptr) {
```
destruct_node всегда вызывается на корень и все что ниже. Казалось бы, на родителя можно никогда не смотреть - все равно после отработки destruct_node ни одной ноды не останется

seshWCS replied:
Убрал

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+                nd->pr->r = nullptr;
+            }
+        }
+        nd->L = nullptr;
```
нет особого смысла выставлять здесь везде nullptr, если nd уничтожвется все равно

seshWCS replied:
Мне мидл из яндекса сказал лучше поставить нуллптр для удобства. Он по совместительству еще мой бывший семинарист по ОиМПу

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+    public:
+        iterator() = default;
+        ~iterator() {
+            nd = nullptr;
```
зачем?

seshWCS replied:
ну просто чтобы было

seshWCS replied:
Ну типо чтобы оно не делитило ноду. Иначе может быть плохо с тем, что ноды будут удаляться при разных операциях итераторами

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+                *this = iterator(nd, root);
+                return ans;
+            }
+            if (nd == root->R) {
```
непонятно, зачем отдельно рассматривать случай с корнем. Семантика итерирования должна быть такой: поднимаемся наверх, пока хоть раз не шагнем вправо

seshWCS replied:
Это нужно просто чтобы определять является ли элемент самым большим из всех в сете. Для инкремента в .end()

adensur commented on Set.h:
```diff
@@ -0,0 +1,443 @@
+    struct node {
+        node* l = nullptr;
+        node* r = nullptr;
+        node* L = nullptr;
```
А зачем нужны L и R?

seshWCS replied:
Это самый левый из потомков и самый правый. Они тут нужны для инкремента иногда и в целом удобная штука

seshWCS committed:
Deleted unused if

seshWCS committed:
Добавил еще 1 иф в инсерт

Это было необязательно т.к. если его бы не было, то в моем дереве все равно был бы корректный инсерт, потому что тогда у меня ни в какой иф ничего не заходило бы и просто возвращалось то же самое дерево, но раз ты оставил коммент, то можно и добавить пару строк по идее.
